shared/rangeanalysis/codeql/rangeanalysis/RangeAnalysis.qll

```diff
@@ -296,11 +373,33 @@ signature module LangSig<Semantic Sem, DeltaSig D> {
 }
 
 signature module BoundSig<LocationSig Location, Semantic Sem, DeltaSig D> {
+  /**
+   * A bound that the range analysis can infer for a variable. This includes
+   * constant bounds represented by the abstract value zero, SSA bounds for when
+   * a variable is bounded by the value of a different variable, and possibly
+   * other abstract values that may be useful variable bounds. Since all bounds
+   * are combined with an integer delta there's no need to represent constant
+   * bounds other than zero.
+   */
   class SemBound {
+    /**
+     * Gets a string representation of this bound.
+     */
     string toString();
 
+    /**
+     * Gets the location of this bound.
+     */
     Location getLocation();
 
+    /**
+     * Gets an expression that equals this bound plus `delta`. For example given
+     * the expression `x = foo() + 1` the variable `x` has a bound with
+     * expression `call to foo()` and delta `-1`.
```

Comment:
I think this example is problematic for a couple of reasons. It's very much not given that `call to foo()` is going to be part of the `Bound` type. In fact I expect it not to. It isn't the case for Java, and I don't _think_ it's the case for C++ either (although I could be wrong there). Secondly, the example is not actually an example for this predicate, but rather an example for one of the predicates inside the range analysis main loop. The example expression is thus never a result of this predicate, and assuming that `call to foo()` is included as an interesting bound, the corresponding tuple (singular!) in this predicate will be the call as the result expression, the bound as `this` and `delta = 0`.

There are really only two canonical cases we can use as examples: SSA bounds and the Zero bound, for which I tried (and perhaps failed) to explain their results in the two lines below.

Another thing to note: Back in December when I worked on refactoring a lot of this, I also looked at making the `Bound` implementation shared, which would likely make this predicate an internal implementation detail. I didn't quite get there due to some inconsistencies in the C++ version, which I didn't manage to sort out before other priorities took over.

Reply:
I have deleted the example and iterated on the rest of the comment, for clarity.

Reply:
I don't see any new changes. Did you forget to push?

Reply:
Er, yes, looks like my push failed due to remote changes.  Sorry.  Pushed properly now.